server/athenian/api/align/mutations/goals.py

```diff
@@ -0,0 +1,129 @@
+            new_goal_id = await _insert_db_goal(parse_result, sdb_conn)
+
+    # TODO: return the complete response
+    return {"errors": None, "goal": {"id": new_goal_id}}
+
+
+class CreateGoalError(Exception):
+    """An error during createGoal mutation handling."""
+
+
+@dataclass(frozen=True)
+class _CreateGoalInputParseResult:
+    goal: Goal
+    team_goals: Sequence[TeamGoal]
+
+
+def _parse_create_goal_input(
+    input: Dict[str, Any],
+    account_id: int,
+) -> _CreateGoalInputParseResult:
+    """Parse CreateGoalInput into Goal and TeamGoal application models."""
+    template_id = int(input["templateId"])
+    if template_id not in TEMPLATES_COLLECTION:
+        raise CreateGoalError(f"Invalid templateId {template_id}")
+
+    team_goals = [_parse_team_goal_input(tg_input) for tg_input in input["teamGoals"]]
+    if not team_goals:
+        raise CreateGoalError("At least one teamGoals is required")
+
+    if len({team_goal.team_id for team_goal in team_goals}) < len(team_goals):
+        raise CreateGoalError("More than one team goal with the same teamId")
+
+    valid_from, expires_at = _convert_goal_dates(input["validFrom"], input["expiresAt"])
+
+    goal = Goal(
+        account_id=account_id,
+        template_id=template_id,
+        valid_from=valid_from,
+        expires_at=expires_at,
+    )
+    return _CreateGoalInputParseResult(goal, team_goals)
+
+
+def _parse_team_goal_input(team_goal_input: dict) -> TeamGoal:
+    """Parse TeamGoalInput into a Team model."""
+    try:
+        # get the first non null value in GoalTargetInput
+        target = next(tgt for tgt in team_goal_input["target"].values() if tgt is not None)
+    except StopIteration:
+        raise CreateGoalError(f'Invalid target for teamId {team_goal_input["teamId"]}')
+
+    return TeamGoal(team_id=team_goal_input["teamId"], target=target)
+
+
+def _convert_goal_dates(valid_from: date, expires_at: date) -> Tuple[datetime, datetime]:
+    """Convert date objects from API into datetimes."""
+    valid_from = datetime.combine(valid_from, time.min, tzinfo=timezone.utc)
+    # expiresAt semantic is to include the given day, so datetime is set to the start of the
+    # following day
+    expires_at = datetime.combine(expires_at + timedelta(days=1), time.min, tzinfo=timezone.utc)
+    return (valid_from, expires_at)
+
+
+async def _parse_result_db_validation(
+    parse_result: _CreateGoalInputParseResult, sdb_conn: Connection,
+) -> None:
+    """Execute validation on parse result using the DB."""
+    # check that all team exist and belong to the right account
+    team_ids = [team_goal.team_id for team_goal in parse_result.team_goals]
+    teams_stmt = sa.select([Team.id]).where(
+        sa.and_(Team.id.in_(team_ids), Team.owner_id == parse_result.goal.account_id),
+    )
+    existing_team_ids_rows = await sdb_conn.fetch_all(teams_stmt)
+    existing_team_ids = sorted(r[0] for r in existing_team_ids_rows)
+
+    if sorted(team_ids) != existing_team_ids:
```

Comment:
We should cast to sets here because the user can send duplicate teams.

Reply:
that was checked in `_parse_create_goal_input`, anyway changed to sets.